README
=============================================
       Google TTS script for Asterisk
=============================================

This script makes use of Google's translate text to speech service
in order to render text to speech and play it back to the user.
It supports a variety of different languages.

This TTS service is 'unofficial' and not supported by Google,
it can be terminated at any point with no warning.
People looking for TTS solutions to base their projects/products on
should look for alternative, officially supported  services.

------------
Requirements
------------
Perl         The Perl Programming Language
perl-libwww  The World-Wide Web library for Perl
perl-Crypt-SSLeay or perl-IO-Socket-SSL for SSL/TLS encryption
sox          Sound eXchange, sound processing program
mpg123       MPEG Audio Player and decoder
Internet access in order to contact google and get the voice data.

------------
Installation
------------
To install copy googletts.agi to your agi-bin directory.
Usually this is /var/lib/asterisk/agi-bin/
To make sure check your /etc/asterisk/asterisk.conf file

-----
Usage
-----
agi(googletts.agi,"text",[language],[intkey],[speed]): This will invoke the Google TTS
engine, render the text string to speech and play it back to the user.
If 'intkey' is set the script will wait for user input. Any given interrupt keys will
cause the playback to immediately terminate and the dialplan to proceed to the
matching extension (this is mainly for use in IVR, see README for examples).
If 'speed' is set the speech rate is altered by that factor (defaults to 1.2).

The script contacts Google's TTS service in order to get the voice data
which then stores in a local cache (by default /tmp/) for future use.
Parameters like default language, enabling or disabling caching and cache dir
can be set up by editing the script.

--------
Examples
--------
sample dialplan code for your extensions.conf

;GoogleTTS Demo
;PLayback messages to user

exten => 1234,1,Answer()
    ;;Play mesage in English:
exten => 1234,n,agi(googletts.agi,"This is a simple google text to speech test in english.",en)
    ;;Play message in Spanish
exten => 1234,n,agi(googletts.agi,"Esta es una simple prueba en español.",es)
    ;;Play message in Greek
exten => 1234,n,agi(googletts.agi,"Αυτό είναι ένα απλό τέστ στα ελληνικά.",el)

;A simple dynamic IVR using GoogleTTS

[my_ivr]
exten => s,1,Answer()
exten => s,n,Set(TIMEOUT(digit)=5)
exten => s,n,agi(googletts.agi,"Welcome to my small interactive voice response menu.",en)
    ;;Wait for digit:
exten => s,n(start),agi(googletts.agi,"Please dial a digit.",en,any)
exten => s,n,WaitExten()

    ;;PLayback the name of the digit and wait for another one:
exten => _X,1,agi(googletts.agi,"You just pressed ${EXTEN}. Try another one please.",en,any)
exten => _X,n,WaitExten()

exten => i,1,agi(googletts.agi,"Invalid extension.",en)
exten => i,n,goto(s,start)

exten => t,1,agi(googletts.agi,"Request timed out.",en)
exten => t,n,goto(s,start)

exten => h,1,Hangup()

-------------------
Supported Languages
-------------------
Afrikaans:          af
Albanian:           sq
Amharic:            am
Arabic:             ar
Armenian:           hy
Azerbaijani:        az
Basque:             eu
Belarusian:         be
Bengali:            bn
Bihari:             bh
Bosnian:            bs
Breton:             br
Bulgarian:          bg
Cambodian:          km
Catalan:            ca
Chinese Simplified: zh-CN
Chinese Traditional:zh-TW
Corsican:           co
Croatian:           hr
Czech:              cs
Danish:             da
Dutch:              nl
English:            en
Esperanto:          eo
Estonian:           et
Faroese:            fo
Filipino:           tl
Finnish:            fi
French:             fr
Frisian:            fy
Galician:           gl
Georgian:           ka
German:             de
Greek:              el
Guarani:            gn
Gujarati:           gu
Hausa:              ha
Hebrew:             iw
Hindi:              hi
Hungarian:          hu
Icelandic:          is
Indonesian:         id
Interlingua:        ia
Irish:              ga
Italian:            it
Japanese:           ja
Javanese:           jw
Kannada:            kn
Kazakh:             kk
Kinyarwanda:        rw
Kirundi:            rn
Korean:             ko
Kurdish:            ku
Kyrgyz:             ky
Laothian:           lo
Latin:              la
Latvian:            lv
Lingala:            ln
Lithuanian:         lt
Macedonian:         mk
Malagasy:           mg
Malay:              ms
Malayalam:          ml
Maltese:            mt
Maori:              mi
Marathi:            mr
Moldavian:          mo
Mongolian:          mn
Montenegrin:        sr-ME
Nepali:             ne
Norwegian:          no
Norwegian Nynorsk:  nn
Occitan:            oc
Oriya:              or
Oromo:              om
Pashto:             ps
Persian:            fa
Polish:             pl
Portuguese:         pt
Portuguese Brazil:  pt-BR
Portuguese Portugal:pt-PT
Punjabi:            pa
Quechua:            qu
Romanian:           ro
Romansh:            rm
Russian:            ru
Scots Gaelic:       gd
Serbian:            sr
Serbo-Croatian:     sh
Sesotho:            st
Shona:              sn
Sindhi:             sd
Sinhalese:          si
Slovak:             sk
Slovenian:          sl
Somali:             so
Spanish:            es
Sundanese:          su
Swahili:            sw
Swedish:            sv
Tajik:              tg
Tamil:              ta
Tatar:              tt
Telugu:             te
Thai:               th
Tigrinya:           ti
Tonga:              to
Turkish:            tr
Turkmen:            tk
Twi:                tw
Uighur:             ug
Ukrainian:          uk
Urdu:               ur
Uzbek:              uz
Vietnamese:         vi
Welsh:              cy
Xhosa:              xh
Yiddish:            yi
Yoruba:             yo
Zulu:               zu

-----------------------
Security Considerations
-----------------------
This script contacts Google servers in order to get voice data.
The script uses TLS to encrypt all the traffic between your pbx and these servers
so no 3rd party can eavesdrop your communication, but your data will be available
to Google under a not yet defined policy.

------------
Tiny version
------------
The '-tiny' suffixed scripts use the HTTP::Tiny perl module instead of LWP::UserAgent.
This makes them a lot faster when run in small embedded systems or boards like
the Raspberry pi. They can be used as an in-place replacement of the normal versions
of the TTS scripts and expose the same interface/cli args. To use them just make sure
you have HTTP::Tiny installed.
In debian or ubuntu based distros: 'apt-get install libhttp-tiny-perl'
In distros that don't have it in their repos: 'cpan HTTP::Tiny'
It also requires mp3 format support in sox in order to avoid the use of mpg123.

-------
License
-------
The GoogleTTS script for asterisk is distributed under the GNU General Public
License v2. See COPYING for details.

--------
Homepage
--------
http://zaf.github.com/asterisk-googletts/


ChangeLog
2015-29-07 Lefteris Zafiris <zaf@fastmail.com> - 0.8
	Compatibility fixes for the updated translate page.
	Set default speed factor to 1. Small changes to cache path handling.
	Fixed UTF handling and URL escaping. Updated text splitting function,
	char limit seems to be 99 for some languages.
	Enabled sox auto-detection by default.
	Added SSL support and made it the default.
	Added 'tiny' versions of the tts scripts.

2012-30-09 Lefteris Zafiris <zaf@fastmail.com> - 0.7
	Added user defined speed factor option from dialplan.
	Sox version auto-detection.	Automatic sound format detection.
	Proxy support. Improved temp file handling.
	Use of 'ConnChace' to speed up http requests.

2012-03-18 Lefteris Zafiris <zaf@fastmail.com> - 0.6
	Added speed factor option. Major rewrite of the cli app.
	Many small fixes.

2012-01-18 Lefteris Zafiris <zaf@fastmail.com> - 0.5
	Removed the waiting for user input after the end of playback.
	It is now advised to use Waitexten() in IVR to get that behavior.
    Updated dialplan examples in README accordingly.
	Replaced POSIX character class [:punct:] in text splitting regural
	expressions. This fixes a misbehavior with some unicode strings.
	Added sample cli app that does speech synthesis.

2011-12-20 Lefteris Zafiris <zaf@fastmail.com> - 0.4
	Added option for verbose debugging messages.
	Fixed interrupt handling and cleaning of temp files.
	Multiline and unicode input is now properly handled.
	Code cleanup.

2011-12-13 Lefteris Zafiris <zaf@fastmail.com> - 0.3
	Fixed compatibility with asterisk 1.4 and older.
	Changed the arguments of tempfile() for compatibility
	with older perl versions.
	Encode ampersands and plus sign to ascii values.
	Added check for channel status before answering.
	When 'intkey' is set the script waits for user input.

2011-12-01 Lefteris Zafiris <zaf@fastmail.com> - 0.2
	Working cache mechanism. Input sanitisation
	and removal of dangerous characters. Better
	error handling. Support for sampling rates
	other than 8KHz. Cleanup mechanism for
	temporary files.

2011-11-28 Lefteris Zafiris <zaf@fastmail.com> - 0.1
	Initial release 0.1


cli/googletts-cli-tiny.pl
#!/usr/bin/env perl

#
# Script that uses Google Translate for text to speech synthesis.
#
# Copyright (C) 2012 - 2015, Lefteris Zafiris <zaf@fastmail.com>
#
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2.
#

use warnings;
use strict;
use utf8;
use Encode qw(decode encode);
use Getopt::Std;
use File::Temp qw(tempfile);
use URI::Escape;
use HTTP::Tiny;

my %options;
my @text;
my @mp3list;
my $samplerate;
my $input;
my $speed   = 1;
my $lang    = "en";
my $tmpdir  = "/tmp";
my $timeout = 10;
my $url     = "https://translate.google.com";
my $mpg123  = `/usr/bin/which mpg123`;
my $sox     = `/usr/bin/which sox`;

VERSION_MESSAGE() if (!@ARGV);

getopts('o:l:r:t:f:s:hqv', \%options);

# Dislpay help messages #
VERSION_MESSAGE() if (defined $options{h});
lang_list("dislpay") if (defined $options{v});

if (!$mpg123 || !$sox) {
	say_msg("mpg123 or sox is missing. Aborting.");
	exit 1;
}
chomp($mpg123, $sox);

parse_options();
$input = decode('utf8', $input);
for ($input) {
	# Split input to comply with google tts requirements #
	s/[\\|*~<>^\n\(\)\[\]\{\}[:cntrl:]]/ /g;
	s/\s+/ /g;
	s/^\s|\s$//g;
	if (!length) {
		say_msg("No text passed for synthesis.");
		exit 1;
	}
	$_ .= "." unless (/^.+[.,?!:;]$/);
	@text = /.{1,150}[.,?!:;]|.{1,150}\s/g;
}
my $lines = @text;

# Initialise User angent #
my $http = HTTP::Tiny->new(
	agent      => 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.101 Safari/537.36',
	timeout    => $timeout,
	verify_SSL => 1,
);

for (my $i = 0; $i < $lines; $i++) {
	# Get speech data from google and save them in temp files #
	my $len = length($text[$i]);
	my $line = encode('utf8', $text[$i]);
	$line =~ s/^\s+|\s+$//g;
	next if (length($line) == 0);
	my $token = make_token($line);
	$line = uri_escape($line);
	my ($mp3_fh, $mp3_name) = tempfile(
		"tts_XXXXXX",
		DIR    => $tmpdir,
		SUFFIX => ".mp3",
		UNLINK => 1
	);
	my $headers = {
		'Accept'          => '*/*',
		'Accept-Encoding' => 'identity;q=1, *;q=0',
		'Accept-Language' => 'en-US,en;q=0.5',
		'DNT'             => '1',
		'Range'           => 'bytes=0-',
		'Referer'         => 'https://translate.google.com/',
	};
	my $response = $http->mirror(
		"$url/translate_tts?ie=UTF-8&q=$line&tl=$lang&total=$lines&idx=$i&textlen=$len&client=tw-ob&tk=$token&prev=input",
		$mp3_name,
		$headers,
	);

	if (!$response->{success}) {
		say_msg("Failed to fetch speech data: ", $response->{status}, $response->{reason});
		exit 1;
	} else {
		push(@mp3list, $mp3_name);
	}
}

# decode mp3s and concatenate #
my ($wav_fh, $wav_name) = tempfile(
	"tts_XXXXXX",
	DIR    => $tmpdir,
	SUFFIX => ".wav",
	UNLINK => 1
);
if (system($mpg123, "-q", "-w", $wav_name, @mp3list)) {
	say_msg("mpg123 failed to process sound file.");
	exit 1;
}

# Set sox args and process wav file #
my @soxargs = ($sox, "-q", $wav_name);
defined $options{o} ? push(@soxargs, ($options{o})) : push(@soxargs, ("-t", "alsa", "-d"));
push(@soxargs, ("tempo", "-s", $speed)) if ($speed != 1);
push(@soxargs, ("rate", $samplerate)) if ($samplerate);

if (system(@soxargs)) {
	say_msg("sox failed to process sound file.");
	exit 1;
}

exit 0;

sub say_msg {
	# Print messages to user if 'quiet' flag is not set #
	my @message = @_;
	warn @message if (!defined $options{q});
	return;
}

sub parse_options {
	# Get input text #
	if (defined $options{t}) {
		$input = $options{t};
	} elsif (defined $options{f}) {
		if (open(my $fh, "<", "$options{f}")) {
			$input = do { local $/; <$fh> };
			close($fh);
		} else {
			say_msg("Cant read file $options{f}");
			exit 1;
		}
	} else {
		say_msg("No text passed for synthesis.");
		exit 1;
	}

	# set the speech language #
	if (defined $options{l}) {
		$options{l} =~ /^[a-zA-Z]{2}(-[a-zA-Z]{2})?$/ ? $lang = $options{l}
			: say_msg("Invalid language setting, using default.");
	}

	# set audio sample rate #
	if (defined $options{r}) {
		$options{r} =~ /\d+/ ? $samplerate = $options{r}
			: say_msg("Invalid sample rate, using default.");
	}

	# set speed factor #
	if (defined $options{s}) {
		$options{s} =~ /\d+/ ? $speed = $options{s}
			: say_msg("Invalind speed factor, using default.");
	}
	return;
}

# Obfuscated crap straight from Google:
# https://translate.google.com/translate/releases/twsfe_w_20151214_RC03/r/js/desktop_module_main.js
sub make_token {
	my $text = shift;
	my $time = int(time() / 3600);
	my @chars = unpack('U*', $text);
	my $stamp = $time;

	foreach (@chars) {
		$stamp = make_rl($stamp + $_, '+-a^+6');
	}
	$stamp = make_rl($stamp, '+-3^+b+-f');
	if ($stamp < 0) {
		$stamp = ($stamp & 2147483647) + 2147483648;
	}
	$stamp %= 10**6;
	return ($stamp . '.' . ($stamp ^ $time));
}

sub make_rl {
	my ($num, $str) = @_;

	for (my $i = 0; $i < length($str) - 2 ; $i += 3) {
		my $d = substr($str, $i+2, 1);
		if (ord($d) >= ord('a')) {
			$d = ord($d) - 87;
		} else {
			$d = int($d);
		}
		if (substr($str, $i+1, 1) eq '+') {
			$d = $num >> $d;
		} else {
			$d = $num << $d;
		}
		if (substr($str, $i, 1) eq '+') {
			$num = $num + $d & 4294967295;
		} else {
			$num = $num ^ $d;
		}
	}
	return $num;
}

sub VERSION_MESSAGE {
	# Help message #
	print "Text to speech synthesis using google voice.\n\n",
		"Supported options:\n",
		" -t <text>      text string to synthesize\n",
		" -f <file>      text file to synthesize\n",
		" -l <lang>      specify the language to use, defaults to 'en' (English)\n",
		" -o <filename>  write output as WAV file\n",
		" -r <rate>      specify the output sampling rate in Hertz (default 22050)\n",
		" -s <factor>    specify the output speed factor\n",
		" -q             quiet (Don't print any messages or warnings)\n",
		" -h             this help message\n",
		" -v             suppoted languages list\n\n",
		"Examples:\n",
		"$0 -l en -t \"Hello world\"\n Have the synthesized speech played back to the user.\n",
		"$0 -o hello.wav -l en -t \"Hello world\"\n Save the synthesized speech as a wav file.\n\n";
	exit 1;
}

sub lang_list {
	# Display the list of supported languages to the user or return it as a hash #
	my $opt      = shift;
	my %sup_lang = (
		"Afrikaans",            "af",    "Albanian",              "sq",    "Amharic",             "am",
		"Arabic",               "ar",    "Armenian",              "hy",    "Azerbaijani",         "az",
		"Basque",               "eu",    "Belarusian",            "be",    "Bengali",             "bn",
		"Bihari",               "bh",    "Bosnian",               "bs",    "Breton",              "br",
		"Bulgarian",            "bg",    "Cambodian",             "km",    "Catalan",             "ca",
		"Chinese (Simplified)", "zh-CN", "Chinese (Traditional)", "zh-TW", "Corsican",            "co",
		"Croatian",             "hr",    "Czech",                 "cs",    "Danish",              "da",
		"Dutch",                "nl",    "English",               "en",    "Esperanto",           "eo",
		"Estonian",             "et",    "Faroese",               "fo",    "Filipino",            "tl",
		"Finnish",              "fi",    "French",                "fr",    "Frisian",             "fy",
		"Galician",             "gl",    "Georgian",              "ka",    "German",              "de",
		"Greek",                "el",    "Guarani",               "gn",    "Gujarati",            "gu",
		"Hausa",                "ha",    "Hebrew",                "iw",    "Hindi",               "hi",
		"Hungarian",            "hu",    "Icelandic",             "is",    "Indonesian",          "id",
		"Interlingua",          "ia",    "Irish",                 "ga",    "Italian",             "it",
		"Japanese",             "ja",    "Javanese",              "jw",    "Kannada",             "kn",
		"Kazakh",               "kk",    "Kinyarwanda",           "rw",    "Kirundi",             "rn",
		"Korean",               "ko",    "Kurdish",               "ku",    "Kyrgyz",              "ky",
		"Laothian",             "lo",    "Latin",                 "la",    "Latvian",             "lv",
		"Lingala",              "ln",    "Lithuanian",            "lt",    "Macedonian",          "mk",
		"Malagasy",             "mg",    "Malay",                 "ms",    "Malayalam",           "ml",
		"Maltese",              "mt",    "Maori",                 "mi",    "Marathi",             "mr",
		"Moldavian",            "mo",    "Mongolian",             "mn",    "Montenegrin",         "sr-ME",
		"Nepali",               "ne",    "Norwegian",             "no",    "Norwegian (Nynorsk)", "nn",
		"Occitan",              "oc",    "Oriya",                 "or",    "Oromo",               "om",
		"Pashto",               "ps",    "Persian",               "fa",    "Polish",              "pl",
		"Portuguese (Brazil)",  "pt-BR", "Portuguese (Portugal)", "pt-PT", "Portuguese",          "pt",
		"Punjabi",              "pa",    "Quechua",               "qu",    "Romanian",            "ro",
		"Romansh",              "rm",    "Russian",               "ru",    "Scots Gaelic",        "gd",
		"Serbian",              "sr",    "Serbo-Croatian",        "sh",    "Sesotho",             "st",
		"Shona",                "sn",    "Sindhi",                "sd",    "Sinhalese",           "si",
		"Slovak",               "sk",    "Slovenian",             "sl",    "Somali",              "so",
		"Spanish",              "es",    "Sundanese",             "su",    "Swahili",             "sw",
		"Swedish",              "sv",    "Tajik",                 "tg",    "Tamil",               "ta",
		"Tatar",                "tt",    "Telugu",                "te",    "Thai",                "th",
		"Tigrinya",             "ti",    "Tonga",                 "to",    "Turkish",             "tr",
		"Turkmen",              "tk",    "Twi",                   "tw",    "Uighur",              "ug",
		"Ukrainian",            "uk",    "Urdu",                  "ur",    "Uzbek",               "uz",
		"Vietnamese",           "vi",    "Welsh",                 "cy",    "Xhosa",               "xh",
		"Yiddish",              "yi",    "Yoruba",                "yo",    "Zulu",                "zu"
	);

	if ($opt eq "dislpay") {
		print "Supported Languages list:\n";
		printf("%-22s:  %s\n", $_, $sup_lang{$_}) foreach (sort keys %sup_lang);
		exit 1;
	} elsif ($opt eq "list") {
		return %sup_lang;
	}
}


cli/googletts-cli.pl
#!/usr/bin/env perl

#
# Script that uses Google Translate for text to speech synthesis.
#
# Copyright (C) 2012 - 2015, Lefteris Zafiris <zaf@fastmail.com>
#
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2.
#

use warnings;
use strict;
use utf8;
use Encode qw(decode encode);
use Getopt::Std;
use File::Temp qw(tempfile);
use URI::Escape;
use LWP::UserAgent;
use LWP::ConnCache;

my %options;
my @text;
my @mp3list;
my $samplerate;
my $input;
my $speed   = 1;
my $lang    = "en";
my $tmpdir  = "/tmp";
my $timeout = 10;
my $url     = "https://translate.google.com";
my $mpg123  = `/usr/bin/which mpg123`;
my $sox     = `/usr/bin/which sox`;

VERSION_MESSAGE() if (!@ARGV);

getopts('o:l:r:t:f:s:hqv', \%options);

# Dislpay help messages #
VERSION_MESSAGE() if (defined $options{h});
lang_list("dislpay") if (defined $options{v});

if (!$mpg123 || !$sox) {
	say_msg("mpg123 or sox is missing. Aborting.");
	exit 1;
}
chomp($mpg123, $sox);

parse_options();
$input = decode('utf8', $input);
for ($input) {
	# Split input to comply with google tts requirements #
	s/[\\|*~<>^\n\(\)\[\]\{\}[:cntrl:]]/ /g;
	s/\s+/ /g;
	s/^\s|\s$//g;
	if (!length) {
		say_msg("No text passed for synthesis.");
		exit 1;
	}
	$_ .= "." unless (/^.+[.,?!:;]$/);
	@text = /.{1,150}[.,?!:;]|.{1,150}\s/g;
}
my $lines = @text;

# Initialise User angent #
my $ua = LWP::UserAgent->new(ssl_opts => { verify_hostname => 1 });
$ua->agent("Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.101 Safari/537.36");
$ua->env_proxy;
$ua->conn_cache(LWP::ConnCache->new());
$ua->timeout($timeout);

for (my $i = 0; $i < $lines; $i++) {
	# Get speech data from google and save them in temp files #
	my $len = length($text[$i]);
	my $line = encode('utf8', $text[$i]);
	$line =~ s/^\s+|\s+$//g;
	next if (length($line) == 0);
	my $token   = make_token($line);
	$line = uri_escape($line);
	my ($mp3_fh, $mp3_name) = tempfile(
		"tts_XXXXXX",
		DIR    => $tmpdir,
		SUFFIX => ".mp3",
		UNLINK => 1
	);

	my $request = HTTP::Request->new('GET' => "$url/translate_tts?ie=UTF-8&q=$line&tl=$lang&total=$lines&idx=$i&textlen=$len&client=tw-ob&tk=$token&prev=input");
	$request->header(
		'Accept'          => '*/*',
		'Accept-Encoding' => 'identity;q=1, *;q=0',
		'Accept-Language' => 'en-US,en;q=0.5',
		'DNT'             => '1',
		'Range'           => 'bytes=0-',
		'Referer'         => 'https://translate.google.com/',
	);
	my $response = $ua->request($request, $mp3_name);

	if (!$response->is_success) {
		say_msg("Failed to fetch speech data: ", $response->code, $response->message);
		exit 1;
	} else {
		push(@mp3list, $mp3_name);
	}
}

# decode mp3s and concatenate #
my ($wav_fh, $wav_name) = tempfile(
	"tts_XXXXXX",
	DIR    => $tmpdir,
	SUFFIX => ".wav",
	UNLINK => 1
);
if (system($mpg123, "-q", "-w", $wav_name, @mp3list)) {
	say_msg("mpg123 failed to process sound file.");
	exit 1;
}

# Set sox args and process wav file #
my @soxargs = ($sox, "-q", $wav_name);
defined $options{o} ? push(@soxargs, ($options{o})) : push(@soxargs, ("-t", "alsa", "-d"));
push(@soxargs, ("tempo", "-s", $speed)) if ($speed != 1);
push(@soxargs, ("rate", $samplerate)) if ($samplerate);

if (system(@soxargs)) {
	say_msg("sox failed to process sound file.");
	exit 1;
}

exit 0;

sub say_msg {
	# Print messages to user if 'quiet' flag is not set #
	my @message = @_;
	warn @message if (!defined $options{q});
	return;
}

sub parse_options {
	# Get input text #
	if (defined $options{t}) {
		$input = $options{t};
	} elsif (defined $options{f}) {
		if (open(my $fh, "<", "$options{f}")) {
			$input = do { local $/; <$fh> };
			close($fh);
		} else {
			say_msg("Cant read file $options{f}");
			exit 1;
		}
	} else {
		say_msg("No text passed for synthesis.");
		exit 1;
	}
	# set the speech language #
	if (defined $options{l}) {
		$options{l} =~ /^[a-zA-Z]{2}(-[a-zA-Z]{2})?$/ ? $lang = $options{l}
			: say_msg("Invalid language setting, using default.");
	}

	# set audio sample rate #
	if (defined $options{r}) {
		$options{r} =~ /\d+/ ? $samplerate = $options{r}
			: say_msg("Invalid sample rate, using default.");
	}

	# set speed factor #
	if (defined $options{s}) {
		$options{s} =~ /\d+/ ? $speed = $options{s}
			: say_msg("Invalind speed factor, using default.");
	}
	return;
}

# Obfuscated crap straight from Google:
# https://translate.google.com/translate/releases/twsfe_w_20151214_RC03/r/js/desktop_module_main.js
sub make_token {
	my $text = shift;
	my $time = int(time() / 3600);
	my @chars = unpack('U*', $text);
	my $stamp = $time;

	foreach (@chars) {
		$stamp = make_rl($stamp + $_, '+-a^+6');
	}
	$stamp = make_rl($stamp, '+-3^+b+-f');
	if ($stamp < 0) {
		$stamp = ($stamp & 2147483647) + 2147483648;
	}
	$stamp %= 10**6;
	return ($stamp . '.' . ($stamp ^ $time));
}

sub make_rl {
	my ($num, $str) = @_;

	for (my $i = 0; $i < length($str) - 2 ; $i += 3) {
		my $d = substr($str, $i+2, 1);
		if (ord($d) >= ord('a')) {
			$d = ord($d) - 87;
		} else {
			$d = int($d);
		}
		if (substr($str, $i+1, 1) eq '+') {
			$d = $num >> $d;
		} else {
			$d = $num << $d;
		}
		if (substr($str, $i, 1) eq '+') {
			$num = $num + $d & 4294967295;
		} else {
			$num = $num ^ $d;
		}
	}
	return $num;
}

sub VERSION_MESSAGE {
	# Help message #
	print "Text to speech synthesis using google voice.\n\n",
		"Supported options:\n",
		" -t <text>      text string to synthesize\n",
		" -f <file>      text file to synthesize\n",
		" -l <lang>      specify the language to use, defaults to 'en' (English)\n",
		" -o <filename>  write output as WAV file\n",
		" -r <rate>      specify the output sampling rate in Hertz (default 22050)\n",
		" -s <factor>    specify the output speed factor\n",
		" -q             quiet (Don't print any messages or warnings)\n",
		" -h             this help message\n",
		" -v             suppoted languages list\n\n",
		"Examples:\n",
		"$0 -l en -t \"Hello world\"\n Have the synthesized speech played back to the user.\n",
		"$0 -o hello.wav -l en -t \"Hello world\"\n Save the synthesized speech as a wav file.\n\n";
	exit 1;
}

sub lang_list {
	# Display the list of supported languages to the user or return it as a hash #
	my $opt      = shift;
	my %sup_lang = (
		"Afrikaans",            "af",    "Albanian",              "sq",    "Amharic",             "am",
		"Arabic",               "ar",    "Armenian",              "hy",    "Azerbaijani",         "az",
		"Basque",               "eu",    "Belarusian",            "be",    "Bengali",             "bn",
		"Bihari",               "bh",    "Bosnian",               "bs",    "Breton",              "br",
		"Bulgarian",            "bg",    "Cambodian",             "km",    "Catalan",             "ca",
		"Chinese (Simplified)", "zh-CN", "Chinese (Traditional)", "zh-TW", "Corsican",            "co",
		"Croatian",             "hr",    "Czech",                 "cs",    "Danish",              "da",
		"Dutch",                "nl",    "English",               "en",    "Esperanto",           "eo",
		"Estonian",             "et",    "Faroese",               "fo",    "Filipino",            "tl",
		"Finnish",              "fi",    "French",                "fr",    "Frisian",             "fy",
		"Galician",             "gl",    "Georgian",              "ka",    "German",              "de",
		"Greek",                "el",    "Guarani",               "gn",    "Gujarati",            "gu",
		"Hausa",                "ha",    "Hebrew",                "iw",    "Hindi",               "hi",
		"Hungarian",            "hu",    "Icelandic",             "is",    "Indonesian",          "id",
		"Interlingua",          "ia",    "Irish",                 "ga",    "Italian",             "it",
		"Japanese",             "ja",    "Javanese",              "jw",    "Kannada",             "kn",
		"Kazakh",               "kk",    "Kinyarwanda",           "rw",    "Kirundi",             "rn",
		"Korean",               "ko",    "Kurdish",               "ku",    "Kyrgyz",              "ky",
		"Laothian",             "lo",    "Latin",                 "la",    "Latvian",             "lv",
		"Lingala",              "ln",    "Lithuanian",            "lt",    "Macedonian",          "mk",
		"Malagasy",             "mg",    "Malay",                 "ms",    "Malayalam",           "ml",
		"Maltese",              "mt",    "Maori",                 "mi",    "Marathi",             "mr",
		"Moldavian",            "mo",    "Mongolian",             "mn",    "Montenegrin",         "sr-ME",
		"Nepali",               "ne",    "Norwegian",             "no",    "Norwegian (Nynorsk)", "nn",
		"Occitan",              "oc",    "Oriya",                 "or",    "Oromo",               "om",
		"Pashto",               "ps",    "Persian",               "fa",    "Polish",              "pl",
		"Portuguese (Brazil)",  "pt-BR", "Portuguese (Portugal)", "pt-PT", "Portuguese",          "pt",
		"Punjabi",              "pa",    "Quechua",               "qu",    "Romanian",            "ro",
		"Romansh",              "rm",    "Russian",               "ru",    "Scots Gaelic",        "gd",
		"Serbian",              "sr",    "Serbo-Croatian",        "sh",    "Sesotho",             "st",
		"Shona",                "sn",    "Sindhi",                "sd",    "Sinhalese",           "si",
		"Slovak",               "sk",    "Slovenian",             "sl",    "Somali",              "so",
		"Spanish",              "es",    "Sundanese",             "su",    "Swahili",             "sw",
		"Swedish",              "sv",    "Tajik",                 "tg",    "Tamil",               "ta",
		"Tatar",                "tt",    "Telugu",                "te",    "Thai",                "th",
		"Tigrinya",             "ti",    "Tonga",                 "to",    "Turkish",             "tr",
		"Turkmen",              "tk",    "Twi",                   "tw",    "Uighur",              "ug",
		"Ukrainian",            "uk",    "Urdu",                  "ur",    "Uzbek",               "uz",
		"Vietnamese",           "vi",    "Welsh",                 "cy",    "Xhosa",               "xh",
		"Yiddish",              "yi",    "Yoruba",                "yo",    "Zulu",                "zu"
	);

	if ($opt eq "dislpay") {
		print "Supported Languages list:\n";
		printf("%-22s:  %s\n", $_, $sup_lang{$_}) foreach (sort keys %sup_lang);
		exit 1;
	} elsif ($opt eq "list") {
		return %sup_lang;
	}
}
